File: src/main/java/in/dragonbra/javasteam/util/event/Event.java
package in.dragonbra.javasteam.util.event;

import java.util.HashSet;

public class Event<T extends EventArgs> {
    protected final HashSet<EventHandler<T>> handlers = new HashSet<>();

    public void addEventHandler(EventHandler<T> handler) {
        synchronized (handlers) {
            handlers.add(handler);
        }
    }

    public void removeEventHandler(EventHandler<T> handler) {
        synchronized (handlers) {
            handlers.remove(handler);
        }
    }

    public void handleEvent(Object sender, T e) {
        synchronized (handlers) {
            for (final EventHandler<T> handler : handlers) {
                handler.handleEvent(sender, e);
            }
        }
    }
}

File: src/main/java/in/dragonbra/javasteam/util/event/Event.kt
package `in`.dragonbra.javasteam.util.event

import java.util.concurrent.CopyOnWriteArrayList

class Event<T : EventArgs> {
    private val handlers = CopyOnWriteArrayList<EventHandler<T>>()

    fun addEventHandler(handler: EventHandler<T>) {
        handlers.add(handler)
    }

    fun removeEventHandler(handler: EventHandler<T>) {
        handlers.remove(handler)
    }

    fun handleEvent(sender: Any, e: T) {
        handlers.forEach { it.handleEvent(sender, e) }
    }
}